feat(sdk-lib-mpc): add implementation of paillier-blum for paillier p…

zahin-mohammad opened the pull request:
- adds implementation for paillier blum proofs
- updates @types/node to match current node version (16.18.46)

TICKET: HSM-102, WP-132

johnoliverdriscoll committed:
feat(sdk-lib-mpc): add implementation of paillier-blum for paillier proof

TICKET: HSM-102

zahin-mohammad committed:
feat(sdk-lib-mpc): add types for paillierBlumProof methods

TICKET: HSM-102

zahin-mohammad committed:
chore(sdk-lib-mpc): use triple =

TICKET: HSM-102

zahin-mohammad committed:
chore(sdk-lib-mpc): rename files to *Proof

TICKET: HSM-102

zahin-mohammad committed:
test(sdk-lib-mpc): add paillierBlumProof tests

TICKET: WP-132

zahin-mohammad commented on modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:
```diff
@@ -0,0 +1,134 @@
+}
+
+// https://en.wikipedia.org/wiki/Jacobi_symbol#Implementation_in_C++
+function jacobi(a, n): bigint {
```
We can move this out to a generic math module in the future.

zahin-mohammad committed:
chore(sdk-lib-mpc): rename files to temp

TICKET: WP-132

zahin-mohammad committed:
chore(sdk-lib-mpc): remove temp files

TICKET: WP-132

zahin-mohammad committed:
chore(sdk-lib-mpc): fix fmt

TICKET: WP-132

zahin-mohammad commented on modules/sdk-lib-mpc/test/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
```
Created this to allow tests import json files for fixtures.

islamaminBitGo commented on modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:
```diff
@@ -0,0 +1,134 @@
+  const y = generateY(N, w);
+  // Prover calculates z_i = y_i ^ d mod N
+  const z = y.map((y_i) => modPow(y_i, d, N));
+  // Prover calculates x_i = y_i ^ 1/4 mod N using [HOC - Fact 2.160]
```
a link would be useful here, you can reference this https://cacr.uwaterloo.ca/hac/about/chap2.pdf

islamaminBitGo replied:
you can change this to -> 
This is calculating the inverse of the function y^4 mod N, i.e. y^(1/4), where N = pq is a blum integer using HOC - Fact 2.160 from https://cacr.uwaterloo.ca/hac/about/chap2.pdf

zahin-mohammad replied:
```
  // This is calculating the inverse of the function y^4 mod N,
  // i.e.y ^ (1 / 4), where N = pq is a blum integer using HOC - Fact 2.160
  // from cacr.uwaterloo.ca / hac / about / chap2.pdf
  // Prover generates y_i.
  const y = generateY(N, w);
  // Prover calculates z_i = y_i ^ d mod N
  const z = y.map((y_i) => modPow(y_i, d, N));
```

like this?

islamaminBitGo commented on modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:
```diff
@@ -0,0 +1,134 @@
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint[]} - set of challenges for N
+ */
+function generateY(N, w): bigint[] {
```
Im not sure where in this function does rejection sampling is done with the set $$Z^{*}_{N}$$ in mind. Need to do gcd(y,N) = 1 to check membership as described in Nothing-up-my-sleeve constructions spec here https://www.zkdocs.com/docs/zkdocs/protocol-primitives/nums/.

zahin-mohammad replied:
cc @johnoliverdriscoll

johnoliverdriscoll replied:
Done

zahin-mohammad committed:
chore(sdk-lib-mpc): add reference to paillierBlum

TICKET: WP-132

islamaminBitGo commented on modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:
```diff
@@ -0,0 +1,137 @@
+    .fill(null)
+    .map((_, i) => {
+      const h = bigIntFromBufferBE(
+        createHmac('sha256', Buffer.from([i]))
```
i think according to https://www.zkdocs.com/docs/zkdocs/protocol-primitives/nums/, we need to use an XOF hash function to produce a digest of the bitlength of N.

islamaminBitGo replied:
maybe something like SHAKE-256

Logicwax replied:
sha3-XOF  (shake-256 mode) is typically used for that, especially when doing RSA keys

johnoliverdriscoll replied:
Updated to use sha256 and separators for each field. However, TypeScript complains about the required parameters to createHash. See code comment.

zahin-mohammad commented:
Rebased off of master.

OttoAllmendinger reviewed (commented):
flush (insufficient context)

zahin-mohammad commented:
Rebased off of master

johnoliverdriscoll committed:
chore(sdk-lib-mpc): use shake256 and sampling rejection for generating quadratic residue

TICKET: WP-132

zahin-mohammad committed:
fix(root): update @types/node

This commit:
- updates @types/node to ^16.18.46
- casts all errors to `any` in tests
- creates a better type def for the keccak module
- updates Buffer.from call sites to make ts happy

TICKET: WP-132

zahin-mohammad committed:
fix(sdk-lib-mpc): test nested specs

TICKET: WP-132

zahin-mohammad committed:
test(sdk-lib-mpc): update range proof fixtures

TICKET: WP-132

zahin-mohammad committed:
test(sdk-lib-mpc): update paillier blum proof fixtures

TICKET: WP-132

zahin-mohammad committed:
test(sdk-lib-mpc): rangeproof -> rangeProof

TICKET: WP-132

OttoAllmendinger reviewed (changes requested):
please summit the changes that are not part of the core change as a separate PR (e.g. https://github.com/BitGo/BitGoJS/pull/3806/commits/aab4f14685df1929dd316014f2b390ee0b19e770)

zahin-mohammad commented:
> please summit the changes that are not part of the core change as a separate PR (e.g. [aab4f14](https://github.com/BitGo/BitGoJS/commit/aab4f14685df1929dd316014f2b390ee0b19e770))

Will do, moving that and a few other changes to seperate prs. This will stay in draft until those are merged.

sachushaji reviewed (commented):
Flushing, Kindly re-request when ready

johnoliverdriscoll committed:
feat(sdk-lib-mpc): add noSmallFactors module

Ticket: HSM-145

johnoliverdriscoll committed:
feat(sdk-lib-mpc): rename nsf.sigma to nsf.rho to match whitepaper

Ticket: HSM-145

zhongxishen committed:
Merge remote-tracking branch 'origin/nsf-proof' into paillier-blum-2

zahin-mohammad commented:
Closing in place of https://github.com/BitGo/BitGoJS/pull/3940
